u_roller.go

```diff
@@ -0,0 +1,110 @@
+package tls
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"net"
+	"sync"
+	"time"
+)
+
+type Roller struct {
+	HelloIDs            []ClientHelloID
+	HelloIDMu           sync.Mutex
+	NextHelloID         ClientHelloID
+	TcpDialTimeout      time.Duration
+	TlsHandshakeTimeout time.Duration
+}
+
+// NewRoller creates Roller object with default range of HelloIDs to cycle through until a
+// working/unblocked one is found.
+func NewRoller() (*Roller, error) {
+	tcpDialTimeoutInc, err := getRandUint16()
+	if err != nil {
+		return nil, err
+	}
+	tcpDialTimeoutInc = tcpDialTimeoutInc % 14
+
+	tlsHandshakeTimeoutInc, err := getRandUint16()
+	if err != nil {
+		return nil, err
+	}
+	tlsHandshakeTimeoutInc = tlsHandshakeTimeoutInc % 20
+
+	r := &Roller{
+		HelloIDs: []ClientHelloID{
+			HelloChrome_Auto,
+			HelloFirefox_Auto,
+			HelloIOS_Auto,
+		},
+		TcpDialTimeout:      time.Second * time.Duration(7+int(tcpDialTimeoutInc)),
+		TlsHandshakeTimeout: time.Second * time.Duration(11+int(tlsHandshakeTimeoutInc)),
+	}
+
+	// choose random clientHelloID to try next
+	nextHelloIdx, err := getRandUint16()
+	if err != nil {
+		return nil, err
+	}
+	r.NextHelloID = r.HelloIDs[int(nextHelloIdx)%len(r.HelloIDs)]
+
+	return r, nil
+}
+
+// Dial attempts to establish connection to given address using different HelloIDs.
+// If a working HelloID is found, it is used again for subsequent Dials.
+// If tcp connection fails or all HelloIDs are tried, returns with last error.
+//
+// Usage examples:
+//    Dial("tcp4", "google.com:443", "google.com")
+//    Dial("tcp", "10.23.144.22:443", "mywebserver.org")
+func (c *Roller) Dial(network, addr, serverName string) (*UConn, error) {
+	remainingHelloIds := make([]ClientHelloID, len(c.HelloIDs))
+	copy(remainingHelloIds, c.HelloIDs)
+
+	c.HelloIDMu.Lock()
+	nextHelloId := c.NextHelloID // keep using same helloID, if it works
+	c.HelloIDMu.Unlock()
+	for {
+		tcpConn, err := net.DialTimeout(network, addr, c.TcpDialTimeout)
+		if err != nil {
+			return nil, err
+		}
+
+		client := UClient(tcpConn, nil, nextHelloId)
+		client.SetSNI(serverName)
+		client.SetDeadline(time.Now().Add(c.TlsHandshakeTimeout))
+		err = client.Handshake()
+		client.SetDeadline(time.Time{}) // unset timeout
+		if err != nil {
+			for i, ID := range remainingHelloIds {
```

Comment:
I can't follow this logic, but I think it's overcomplicated.  If you shuffle the array when it's created, then you can just walk down the array in order here.

Reply:
Not quite: I start with `NextHelloID`, such that we don't change IDs, when not necessary and stick to a working one while it works.